### ISSUE TITLE
improve compatibility with bioformats

### ISSUE THREAD

**Issue by jcupitt** [libvips/libvips#1365]
See:

https://github.com/libvips/libvips/issues/1253

https://github.com/openmicroscopy/bioformats/issues/3300

https://trello.com/c/pW0zbhAK/306-detect-tiff-pyramids-generated-by-vips

**Comment by Zhaoyang-XU** [comment]
Hi John, is there any progress on this issue?

**Comment by Zhaoyang-XU** [comment]
> Sorry, no. Perhaps next year (unless someone volunteers before then).

Thanks for the updates!

**Comment by gil-c** [comment]
Hello John, any news about BioFormats compatibility? That would be really useful for us.

**Comment by jcupitt** [comment]
Hi, perhaps later in the 8.10 cycle. Right now we're still revising the loaders for the new streaming system.

Very happy to accept a PR of course, if you're able.

**Comment by gil-c** [comment]
I'll try to help as can.

I downloaded the following ome.tiff image: https://downloads.openmicroscopy.org/images/OME-TIFF/2016-06/sub-resolutions/Fluorescence/LuCa-7color_Scan1.ome.tiff

I open it in Python with: im = vips.Image.tiffload("LuCa-7color_Scan1.ome.tiff", n=5)

It turns out that in this file, the 5 first pages have the same size, be not the following 3. So I cannot load the 8 pages at once.

Then I saved the image like this: im.tiffsave("LuCa-7color_Scan1_libvips.tiff", compression="lzw", tile=True, tile_width=512, tile_height=512, page_height=34560)

And I compared the original and the new new file with tiffinfo. This is the second page from the original file (I skipped the first page cause it contains a very long ImageDescription):

TIFF Directory at offset 0x14d40236 (349438518)
  Image Width: 24960 Image Length: 34560
  Tile Width: 512 Tile Length: 512
  Resolution: 20080.5, 20080.5 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: LZW
  Photometric Interpretation: min-is-black
  Samples/Pixel: 1
  Planar Configuration: single image plane
  SubIFD Offsets: 1619859285 1943702908 2028625295 2054020223 2060864623
  ImageDescription: ImageJ=
hyperstack=true
images=5
channels=5
slices=1
frames=1
  Software: OME Bio-Formats 6.0.0-rc1

And this is what I get from the file saved by libvips:

TIFF Directory at offset 0x1c968e04 (479628804)
  Image Width: 24960 Image Length: 34560
  Tile Width: 512 Tile Length: 512
  Resolution: 20080.5, 20080.5 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: LZW
  Photometric Interpretation: min-is-black
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 1
  Planar Configuration: single image plane
  ImageDescription: ImageJ=
hyperstack=true
images=5
channels=5
slices=1
frames=1
  Predictor: horizontal differencing 2 (0x2)

The main difference I see is the missing "SubIFD Offsets" tag, and I assume the actual data are probably missing too.

Turning the pyramid flag in tiffsave to True doesn't change anything.

Does this help?

**Comment by jcupitt** [comment]
Thank you for testing this @gil-c 

We have some progress on this issue. This branch:

https://github.com/libvips/libvips/tree/ome-tiff-pyramid

Has a tiff writer which can (optionally) write OME-style TIFF pyramids (I think).

You can use it like this:

```
john@kiwi:~/pics$ vips copy k2.jpg x.tif[pyramid,subifd]
john@kiwi:~/pics$ tiffinfo x.tif 
TIFF Directory at offset 0x900008 (9437192)
  Image Width: 1450 Image Length: 2048
  Tile Width: 128 Tile Length: 128
  Resolution: 72.009, 72.009 pixels/inch
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: RGB color
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 3
  Planar Configuration: single image plane
  SubIFD Offsets: 11797894 12388226 12585126 12634522
```

The `subifd` flag means "put pyramid layers into TIFF subdirectories rather than separate pages". As you can see the `SubIFD Offsets` tags points to four other images, each of which has the reduced-resolution tag set.

It can write many-page images too:

```
john@kiwi:~/pics$ vips copy nipguide.pdf[n=2] x.tif[pyramid]
john@kiwi:~/pics$ tiffinfo x.tif 
TIFF Directory at offset 0x230008 (2293768)
  Subfile Type: multi-page document (2 = 0x2)
  Image Width: 595 Image Length: 842
  Tile Width: 128 Tile Length: 128
  Resolution: 28.3465, 28.3465 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: RGB color
  Extra Samples: 1<unassoc-alpha>
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 4
  Planar Configuration: single image plane
  Page Number: 0-2
  SubIFD Offsets: 3080770 3343256 3409070
TIFF Directory at offset 0x5705a4 (5703076)
  Subfile Type: multi-page document (2 = 0x2)
  Image Width: 595 Image Length: 842
  Tile Width: 128 Tile Length: 128
  Resolution: 28.3465, 28.3465 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: RGB color
  Extra Samples: 1<unassoc-alpha>
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 4
  Planar Configuration: single image plane
  Page Number: 1-2
  SubIFD Offsets: 6490078 6752564 6818378
```

Now this has created a two page image, and each page contains a pyramid. The `subifd` mode is automatically enabled in this case. 

It should work for OME tiff as well, as long as each page has the same size. For example:

```
$ vips copy LuCa-7color_Scan1.ome.tiff[n=5] x.tif[pyramid,compression=jpeg]
```

Any testing would be very welcome!

We plan to add a flag to the tiff loader to let libvips read these subifd images.

**Comment by jcupitt** [comment]
It seems to work with bioformats as well:

```
$ vipsthumbnail LuCa-7color_Scan1.ome.tiff[n=5] -s 1000 -o x.tif[pyramid] 
```

Make a five page tiff image, each page has one band, and each page is a pyramid:

```
$ tiffinfo x.tif
TIFF Directory at offset 0xc0008 (786440)
  Subfile Type: multi-page document (2 = 0x2)
  Image Width: 722 Image Length: 1000
  Tile Width: 128 Tile Length: 128
  Resolution: 20080.5, 20080.5 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: min-is-black
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 1
  Planar Configuration: single image plane
  Page Number: 0-5
  SubIFD Offsets: 983672 1049580 1066288
  ImageDescription: ImageJ=
hyperstack=true
images=5
channels=5
slices=1
frames=1
TIFF Directory at offset 0x1c465c (1853020)
  Subfile Type: multi-page document (2 = 0x2)
...
```

And bioformats can read it:

```
$ ./bftools/showinf x.tif 
Checking file format [Tagged Image File Format]
Initializing reader
TiffDelegateReader initializing x.tif
Reading IFDs
Populating metadata
Checking comment style
Populating OME metadata
Initialization took 0.152s

Reading core metadata
filename = x.tif
Series count = 1
Series #0 :
	Image count = 5
	RGB = false (1) 
	Interleaved = false
	Indexed = false (false color)
	Width = 722
	Height = 1000
	SizeZ = 1
	SizeT = 1
	SizeC = 5
	Tile size = 128 x 128
	Thumbnail size = 92 x 128
	Endianness = intel (little)
	Dimension order = XYCZT (uncertain)
	Pixel type = uint8
	Valid bits per pixel = 8
	Metadata complete = true
	Thumbnail series = false
	-----
	Plane #0 <=> Z 0, C 0, T 0
	Plane #2 <=> Z 0, C 2, T 0
	Plane #4 <=> Z 0, C 4, T 0


Reading pixel data (0-4)
	Read 5/5 planes (100%)
[done]
```

I'll have a go at adding read support for subifd as well.

**Comment by petebankhead** [comment]
@jcupitt this sounds great! If you might want to test is interactively across resolutions, QuPath should have full support for OME-TIFF via Bio-Formats ([latest release here](https://github.com/qupath/qupath/releases/tag/v0.2.0-m12)).

**Comment by jcupitt** [comment]
I made a test win build here:

http://www.rollthepotato.net/~john/vips-dev-w64-web-8.10.0.zip

And I'm testing with this image:

https://downloads.openmicroscopy.org/images/OME-TIFF/2016-06/sub-resolutions/Fluorescence/LuCa-7color_Scan1.ome.tiff

With:

```
vipsthumbnail LuCa-7color_Scan1.ome.tiff[n=5] -s 1000 -o x.tif[pyramid] 
```

ie. make a 1000 pixel across version of the first 5 pages (the other two are smaller images).

It loads into qupath, but there seems to be some issue with pixel layout. I'll poke at it a bit more.

It seems to sort-of work, but the images don't load

**Comment by gil-c** [comment]
Hi John,

I was able to use tiffsave with the python bindings and the flags: pyramid=True, subifd=True. It creates the SubIFD properly (visible by tiffinfo). 

Same thing for me, this is not sufficient for QuPath to see the pyramids. Now I am playing with the metadata. BTW, I see that setting them with the 'image-description' attribute results in having them in all Tiff Directories, but I don't think this is an issue for bio-formats. 

I will try to shape the ome.xml content in the proper way and I'll let you know.

**Comment by gil-c** [comment]
🎉 It works!!

This is what I did:

In a terminal and text editor:

1. I extracted the ome.xml metadata from the LuCa-7color_Scan1.ome.tiff file with libtiff,
2. I edited the xml content in notepad++ (I used the plugin XML Tools to format it properly),
3. I removed the section related to the thumbnails that we do not handle with libvips (pages 6-8), 
4. I replaced a strange character that was supposed to be 'µ' but was '┬╡' (I put 'n' instead),
5. I saved the modified metadata in: "LuCa-7color_Scan1_modified.ome.xml"

In Python:
```
im = pyvips.Image.tiffload("LuCa-7color_Scan1.ome.tiff", n=5)
metadata = open("LuCa-7color_Scan1_modified.ome.xml").read()
im.set_type(pyvips.GValue.gstr_type, "image-description", metadata)
im.tiffsave("LuCa-7color_Scan1_from_libvips.ome.tiff", compression="lzw", tile=True, tile_width=512, tile_height=512, page_height=34560, pyramid=True, subifd=True)
```

Now QuPath open this file without issue, and it sees the pyramids!

This is super nice. Now I will try with different images, it should also work fine I think.

**Comment by jcupitt** [comment]
Great!

I'm seeing stripe problems, eg.:

```
vipsthumbnail LuCa-7color_Scan1.ome.tiff[n=5] -s 1000 -o x.tif[pyramid,tile-width=512,tile-height=512,compression=lzw] 
vips copy x.tif[page=1] x.png
```

Makes an image with horizontal strips, looks like they are half a tile high. I'll investigate.

I've added a thing to the tiff loader to let you pick a subifd:

```
vips copy LuCa-7color_Scan1.ome.tiff[page=1,subifd=2] x.png
```

Reads subimage 2 from page 1.

**Comment by jcupitt** [comment]
This branch can now load and save these pyramids correctly (I think). This works:

```
vipsthumbnail LuCa-7color_Scan1.ome.tiff[n=4] -s 1000 -o x.tif[compression=lzw,subifd,pyramid]
```

And the result loads into QuPath. I'll make a test build for Windows.

**Comment by jcupitt** [comment]
Windows test build:

http://www.rollthepotato.net/~john/vips-dev-w64-web-8.10.0a.zip

For linux, build this branch:

https://github.com/libvips/libvips/tree/ome-tiff-pyramid

I'll update the docs and add some tests.

**Comment by jcupitt** [comment]
OK, tests and docs, another win build. I think this is pretty much done now. Any testing very welcome.

http://www.rollthepotato.net/~john/vips-dev-w64-web-8.10.0b.zip

It fails for very small OME thumbnails for some reason, but otherwise seems to work well.

**Comment by jcupitt** [comment]
Here's a complex example:

```
$ vipsthumbnail LuCa-7color_Scan1.ome.tiff[n=5] -s 500 -o x.tif[pyramid,subifd,compression=lzw,tile] --vips-info
VIPS-INFO: 12:16:49.326: thumbnailing LuCa-7color_Scan1.ome.tiff[n=5]
VIPS-INFO: 12:16:49.328: selected loader is VipsForeignLoadTiffFile
VIPS-INFO: 12:16:49.328: input size is 24960 x 172800
VIPS-INFO: 12:16:49.337: loading with factor 4 pre-shrink
VIPS-INFO: 12:16:49.337: pre-shrunk size is 780 x 5400
VIPS-INFO: 12:16:49.337: converting to processing space b-w
VIPS-INFO: 12:16:49.337: residual reducev by 0.462963
VIPS-INFO: 12:16:49.337: reducev: 13 point mask
VIPS-INFO: 12:16:49.339: reducev: using vector path
VIPS-INFO: 12:16:49.339: residual reduceh by 0.462963
VIPS-INFO: 12:16:49.339: reduceh: 13 point mask
VIPS-INFO: 12:16:49.339: thumbnailing LuCa-7color_Scan1.ome.tiff[n=5] as ./x.tif[pyramid,subifd,compression=lzw,tile]
```

So that's thumbnailing a complete OME TIFF pyramid. You can see it's picking out the right layer from the source pyramid and then writing a new pyramid in the same format. The result seems to load into QuPath (for me).

It seems to work fine for larger sizes (500 pixels across here), but fails for small dimensions, eg. 100 pixels. I'll have a look. Otherwise, I don't know of any remaining issues.

**Comment by petebankhead** [comment]
This sounds excellent! The reason I started following this discussion is that I wanted to use the fantastic speed of VIPS to convert .ndpi files (or others supported by OpenSlide) to TIFF for compatibility with OMERO.

And the reason I wanted to convert is because I wanted to strip out any potentially-unwanted metadata (original slide names, labels) so they could be used for teaching.

Sorry I haven't had a chance to build this yet with OpenSlide, but an example image would be OS-1/2/3.ndpi from the OpenSlide test data http://openslide.cs.cmu.edu/download/openslide-testdata/Hamamatsu/

Converting one of these files with Bio-Formats was unusably slow. Converting them with VIPS previously was irresistibly fast, but the output could only be read using OpenSlide - and not Bio-Formats. This was fine for QuPath, but not for OMERO. 

> Sidenote: when you open an image in QuPath it might use OpenSlide and it might use Bio-Formats - it will try both. You can check which was used by checking the *Server type* entry under the *Image* tab.
> 
> You can control this either via setting the *Preference* to set file extensions for Bio-Formats, or by adding your image to a [project (when you can manually select the image provider you want)](https://qupath.readthedocs.io/en/latest/docs/tutorials/projects.html#add-images).

**Comment by jcupitt** [comment]
I think it should work with ndpi, but someone needs to test it. It might need some of the OME metadata added too.

**Comment by gil-c** [comment]
I tried to convert OS-2.ndpi to ome.tif. I found two issues (not related to this post):

1. I got the error: "failed to allocate 28085059584 bytes". Apparently Libvips needs to load the whole image in RAM for ndpi files. This does not occur when the source image is a tif file.
2. Displaying this image in RGB looks different than the QuPath/OpenSlide rendering where channels are associated with specific colors for H&E.

Apart this, the pyramid is now recognized by QuPath/Bio-Formats. Here is the python code I used (I had to pick a lower resolution to overcome issue n°1)

```
# read ndpi file
im = vips.Image.new_from_file("OS-2.ndpi", page=3, access=vips.Access.SEQUENTIAL)

# re-organize in a single band toilet roll image
im2 = im.extract_band(0)
for i in range(1, im.get("bands")):
    im2 = im2.insert(im.extract_band(i), 0, im2.height, expand=True)

# build minimal metadata
metadata = f"""<?xml version="1.0" encoding="UTF-8"?>
<OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2016-06"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd">
    <Image ID="Image:0">
        <!-- Minimum required fields about image dimensions -->
        <Pixels DimensionOrder="XYCZT"
                ID="Pixels:0"
                SizeC="{im.get("bands")}"
                SizeT="1"
                SizeX="{im.width}"
                SizeY="{im.height}"
                SizeZ="1"
                Type="uint16">
        </Pixels>
    </Image>
</OME>"""

# set metadata and save
im2.set_type(vips.GValue.gstr_type, "image-description", metadata)
im2.tiffsave("OS-1.ome.tiff", compression="lzw", tile=True, tile_width=512, tile_height=512, page_height=im.height, pyramid=True, subifd=True)
```

**Comment by gil-c** [comment]
@jcupitt To me everything looks good!

Btw:

1. For some reason (probably character encoding), it is not possible to read the image-description field from the original LuCa-7color_Scan1.ome.tiff file. But once we set a valid ome.xml content without exotic characters, it works (writing and reading).
2. There is no direct method to get the number of IFDs via Libvips, but we can guess from the native resolution, or by parsing the ome metadata. I don't think this is critical.
3. The image-description content is written several times, once in every TIFF Directory. We can see it with tiffinfo. But it doesn't affect the reading by Bio-Formats, so not problematic either.

I did my tests only from python with the tiffload and tiffsave methods. Actually I don't fully understand how you do it with the thumbnail command.

Thank you very much for this nice new feature, this is great!

**Comment by jcupitt** [comment]
Hey, thanks for testing!

Here's a slightly modified version of your convert program:

```python
#!/usr/bin/python3
  
import sys
import pyvips

im = pyvips.Image.new_from_file(sys.argv[1])

# openslide will add an alpha
if im.hasalpha():
    im = im[:-1]

# reorganise as a toilet-roll image
im = pyvips.Image.arrayjoin(im.bandsplit(), across=1)

# before we can modify an image (set metadata in this case), we must take a 
# private copy
im = im.copy()

im.set_type(pyvips.GValue.gint_type, "page-height", im.height)

# set minimal OME metadata
im.set_type(pyvips.GValue.gstr_type, "image-description",
f"""<?xml version="1.0" encoding="UTF-8"?>
<OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2016-06"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd">
    <Image ID="Image:0">
        <!-- Minimum required fields about image dimensions -->
        <Pixels DimensionOrder="XYCZT"
                ID="Pixels:0"
                SizeC="{im.bands}"
                SizeT="1"
                SizeX="{im.width}"
                SizeY="{im.height}"
                SizeZ="1"
                Type="uint16">
        </Pixels>
    </Image>
</OME>""")

im.tiffsave(sys.argv[2], compression="jpeg", tile=True,
            tile_width=512, tile_height=512,
            pyramid=True, subifd=True)
```

1. That Windows binary above does not have openslide, hence your out of memory error. It might work if you leave off the `access=` since it'll load as a TIFF.
2. Perhaps the RGB difference can be fixed in the metadata.
2a. You can get the number of subifds from a TIFF with `im.get("n-subifds")`.

Other notes:

1. I've set it saving with JPG compression to save some disc space.
2. I'll have a look at the char encoding issue.

**Comment by jcupitt** [comment]
Thanks for the minimal OME metadata example, by the way!

Here's a converter for OME TIFFs:

```
#!/usr/bin/python3

import sys
import pyvips

im = pyvips.Image.new_from_file(sys.argv[1])

# OME TIFFs load with bands in pages
ome_height = im.get("page-height")
ome_bands = im.height // ome_height

# set minimal OME metadata
im = im.copy()
im.set_type(pyvips.GValue.gstr_type, "image-description", 
f"""<?xml version="1.0" encoding="UTF-8"?>
<OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2016-06"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd">
    <Image ID="Image:0">
        <!-- Minimum required fields about image dimensions -->
        <Pixels DimensionOrder="XYCZT"
                ID="Pixels:0"
                SizeC="{ome_bands}"
                SizeT="1"
                SizeX="{im.width}"
                SizeY="{ome_height}"
                SizeZ="1"
                Type="uint8">
        </Pixels>
    </Image>
</OME>""")

im.tiffsave(sys.argv[2], compression="jpeg", tile=True, 
            tile_width=512, tile_height=512, 
            pyramid=True, subifd=True)
```

Run with eg.:

```
./convert.py LuCa-7color_Scan1.ome.tiff[n=5] x.tif
```

**Comment by jcupitt** [comment]
Updated Windows build, including openslide support and a couple of other small fixes:

http://www.rollthepotato.net/~john/vips-dev-w64-all-8.10.0c.zip

**Comment by gil-c** [comment]
This build works fine with the ndpi file (no more memory issue).

And `im.get("n-subifds")` works fine (good to have it).

However, the file you posted http://www.rollthepotato.net/~john/OS-2.tif isn't Bio-Formats compatible. There is a mismatch in the image height and the ome height in the metadata.

I used a slightly different python code cause there was no "page-height" field available with OS-2.ndpi.

```
#!/usr/bin/python3

import sys
import pyvips

im = pyvips.Image.new_from_file(sys.argv[1])

# openslide will add an alpha
if im.hasalpha():
    im = im[:-1]

# read useful info
height = im.height
width = im.width
nb_bands = im.bands

# reorganise as a toilet-roll image
im = pyvips.Image.arrayjoin(im.bandsplit(), across=1)

# set minimal OME metadata
im = im.copy()
im.set_type(pyvips.GValue.gstr_type, "image-description",
f"""<?xml version="1.0" encoding="UTF-8"?>
<OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2016-06"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd">
    <Image ID="Image:0">
        <!-- Minimum required fields about image dimensions -->
        <Pixels DimensionOrder="XYCZT"
                ID="Pixels:0"
                SizeC="{nb_bands}"
                SizeT="1"
                SizeX="{width}"
                SizeY="{height}"
                SizeZ="1"
                Type="uint8">
        </Pixels>
    </Image>
</OME>""")

im.tiffsave(sys.argv[2], compression="jpeg", tile=True,
            tile_width=512, tile_height=512,
            pyramid=True, subifd=True, bigtiff=True, page_height=height)
```

QuPath/BioFormats can read the output file, but there is still this difference in color representation. @petebankhead will probably know how to deal with it. 

Yes, the OME metadata has optional fields to set the name and color of each channel, and QuPath reads them correctly. For instance:

```
<?xml version="1.0" encoding="UTF-8"?>
<OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2016-06"
   	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
   	xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd">
	<Image ID="Image:0">
		<!-- Minimum required fields about image dimensions -->
		<Pixels DimensionOrder="XYCZT"
      			ID="Pixels:0"
      			SizeC="5"
      			SizeT="1"
      			SizeX="24960"
      			SizeY="34560"
      			SizeZ="1"
      			Type="uint8">
                       <!-- Optional fields for each channel -->
                       <Channel Color="65535"
       				ID="Channel:0:0"
       				Name="DAPI"/>
			<Channel Color="16711935"
       				ID="Channel:0:1"
       				Name="FITC"/>
			<Channel Color="-65281"
       				ID="Channel:0:2"
       				Name="CY3"/>
			<Channel Color="-8388353"
       				ID="Channel:0:3"
       				Name="Texas Red"/>
			<Channel Color="-16776961"
       				ID="Channel:0:4"
       				Name="CY5"/>
		</Pixels>
	</Image>
</OME>
```

This one comes from the file LuCa-7color_Scan1.ome.tiff

**Comment by jcupitt** [comment]
Argh! You're right, I messed that up badly. I've fixed my python above and uploaded an updated TIFF file. Thanks!

**Comment by petebankhead** [comment]
> QuPath/BioFormats can read the output file, but there is still this difference in color representation. @petebankhead will probably know how to deal with it.

This seems to be behave Bio-Formats hasn't recognized the converted TIFF as an RGB image, which results in QuPath applying some auto brightness/contrast logic. In any case, pixel values should be unaffected and if you reset the brightness/contrast in QuPath for all channels then it should be 'fixed'.

I *think* the RGB interpretation is being made here in the Bio-Formats reader: https://github.com/ome/bioformats/blob/df5c3c04bde5b1322c589b3d902d94f96a575e9f/components/formats-bsd/src/loci/formats/in/MinimalTiffReader.java#L607

Unless there are already vips options that could resolve this, I'm not sure exactly where it is best addressed. Since it isn't critical to reading the image, QuPath probably ought to be a bit less reliant on Bio-Formats recognizing an image as RGB.

Via QuPath's script editor, a quick check to see if an image has been recognized as RGB is with
```groovy
println getCurrentServer().isRGB()
```
and a slightly more informative script that goes straight to Bio-Formats is
```groovy
def path = '/Users/pete/Downloads/OS-2.tif'
try (reader = new loci.formats.ImageReader()) {
    reader.setId(path)
    // Get the underlying (most specific) reader used
    def reader2 = reader.getReader()
    println reader2.isRGB()
}
```
From this I see that Bio-Formats does log a few warnings when opening OS-2.tif:
```
WARN: SamplesPerPixel mismatch: OME=-1, TIFF=1
WARN: Image ID 'Image:0': missing plane #0
WARN: Image ID 'Image:0': missing plane #1
WARN: Image ID 'Image:0': missing plane #2
WARN: Using TiffReader to determine the number of planes.
```

**Comment by jcupitt** [comment]
Hi @petebankhead, there was a dumb problem with my XML. 

Have you tried the updated image? It loads into QuPath for me and appears to be an RGB image.

**Comment by jcupitt** [comment]
Or at least pixels have colours heh. I see:

![Screenshot from 2020-06-05 10-41-31](https://user-images.githubusercontent.com/580843/83861990-3b5a0e00-a719-11ea-8fc5-2c20f67a2f0b.png)

**Comment by petebankhead** [comment]
Hi @jcupitt, I think I got the right image from http://www.rollthepotato.net/~john/OS-2.tif (downloaded just now).

You're right that the channels/colors are all fine - QuPath just handles images that is recognizes as 'inherently' RGB a bit differently, using an ARGB packed representation that may enable some optimizations. In this case, it relies on Bio-Formats returning `reader.isRGB() == true`.

QuPath should be able to quite happily handle images of various bit depths with > 40 channels. If the image isn't flagged as RGB then it is treated as a 3-channel image that may or may not happen to be using RGB lookup tables, but these lookup tables (and channel names) can be freely changed.

I think all this QuPath-specific stuff can be ignored, it really just explains why the colors may differ.

The most relevant bit is that, using Bio-Formats directly (ignoring QuPath), `reader.isRGB() == true` in the original NDPI file and `reader.isRGB() == false` in the converted TIFF. I'm not sure if that matters.

**Comment by jcupitt** [comment]
Ah gotcha. Yes, libvips is saving as a set of three interleaved (min-is-black) one-band images and relying on the XML in IMAGEDESCRIPTION to hint that it's really an RGB image. It sounds like bioformats would prefer the image to be tagged as plane separate RGB.

libvips can read plane separate TIFFs (though it's a bit slow), but has no support right now for writing them. I suppose there would need to be (yet another) writer flag.

**Comment by jcupitt** [comment]
Also, thank you very much for checking this.

**Comment by petebankhead** [comment]
@jcupitt Thank you very much for doing this! On a gentler week I'd have tried to be more proactively helpful and learn to build libvips myself...

Once it is in a release I'll add instructions of the QuPath docs for how to do the conversion. I expect I will use libvips more and more as I learn what it can do.

**Comment by jcupitt** [comment]
PR for this branch: https://github.com/libvips/libvips/pull/1673

I'll close this issue and merge to master so we can start fuzzing. This feature should be in 8.10. Any more problems, please open a new issue.

Thank you everyone!
